{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8ad43c13-6cb6-43ba-90b5-4688aa087375",
   "metadata": {},
   "source": [
    "# Example: Eigendecomposition of Stoichiometric Matrices\n",
    "In this example, we'll explore the eigendecomposition of a standard matrix in chemical systems: the stoichiometric matrix $\\mathbf{S}$. The stoichiometric matrix is the digital representation of a chemical reaction system.\n",
    "\n",
    "> __Learning Objectives:__\n",
    "> \n",
    "> By the end of this example, you should be able to:\n",
    ">\n",
    "> * __Compute covariance matrices and eigendecompositions:__ Calculate the covariance matrix of a stoichiometric matrix and compute its eigenvalues and eigenvectors using both direct and iterative methods.\n",
    "> * __Implement and verify power iteration:__ Apply the power iteration method to find the largest eigenvalue and eigenvector of a symmetric matrix and validate results against built-in linear algebra functions.\n",
    "> * __Interpret eigenvector components for network analysis:__ Analyze eigenvector coefficients to identify important reactions in metabolic networks using scaling transformations.\n",
    "\n",
    "Let's get started!\n",
    "___\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc899abe",
   "metadata": {},
   "source": [
    "## Background: What is a stoichiometric matrix?\n",
    "Suppose we have a set of chemical (or biochemical) reactions $\\mathcal{R}$ involving the chemical species (metabolite) set $\\mathcal{M}$. Then, the stoichiometric matrix is a $\\mathbf{S}\\in\\mathbb{R}^{|\\mathcal{M}|\\times|\\mathcal{R}|}$ matrix that holds the stoichiometric coefficients $\\sigma_{ij}\\in\\mathbf{S}$ such that:\n",
    "* $\\sigma_{ij}>0$: Chemical species (metabolite) $i$ is _produced_ by reaction $j$. Species $i$ is a product of reaction $j$.\n",
    "* $\\sigma_{ij} = 0$: Chemical species (metabolite) $i$ is not connected with reaction $j$\n",
    "* $\\sigma_{ij}<0$: Chemical species (metabolite) $i$ is _consumed_ by reaction $j$. Species $i$ is a reactant of reaction $j$.\n",
    "\n",
    "Thus, the stoichiometric matrix $\\mathbf{S}$ encodes the complete connectivity information of the chemical reaction system for, in this case, a biochemical reaction network. Thus, it is the digital representation of the reaction network inside of a cell! Pretty cool, right?\n",
    "___"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1018fc3-c3b2-4c66-9c03-17e4b0a1d8fa",
   "metadata": {},
   "source": [
    "## Setup, Data, and Prerequisites\n",
    "First, we set up the computational environment by including the `Include.jl` file and loading any needed resources.\n",
    "\n",
    "> The [`include(...)` command](https://docs.julialang.org/en/v1/base/base/#include) evaluates the contents of the input source file, `Include.jl`, in the notebook's global scope. The `Include.jl` file sets paths, loads required external packages, etc. For additional information on functions and types used in this material, see the [Julia programming language documentation](https://docs.julialang.org/en/v1/). \n",
    "\n",
    "Let's set up our code environment:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a955d0e5-44a6-43c2-9e19-690e523d2068",
   "metadata": {},
   "outputs": [],
   "source": [
    "include(joinpath(@__DIR__, \"Include.jl\")); # include the Include.jl file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf89812b",
   "metadata": {},
   "source": [
    "In addition to standard Julia libraries, we'll also use [the `VLDataScienceMachineLearningPackage.jl` package](https://github.com/varnerlab/VLDataScienceMachineLearningPackage.jl). Check out [the documentation](https://varnerlab.github.io/VLDataScienceMachineLearningPackage.jl/dev/) for more information on the functions, types, and data used in this material."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e164a9e-f226-4120-a7cf-5575f73296bf",
   "metadata": {},
   "source": [
    "### Data\n",
    "We developed a simple software development kit (SDK) against [the BiGG Models application programming interface at the University of California, San Diego](http://bigg.ucsd.edu/). The [BiGG Models database](http://bigg.ucsd.edu/) integrates published genome-scale metabolic networks into a single database with standardized nomenclature and structure. \n",
    "\n",
    "> __What are we doing here?__\n",
    "> \n",
    "> We are going to download a stoichiometric matrix from [the BiGG models database](http://bigg.ucsd.edu/) using [the BiGG models API](http://bigg.ucsd.edu/data_access) and then compute its eigendecomposition. \n",
    "> * [The BiGG models API](http://bigg.ucsd.edu/data_access) allows users to programmatically access genome-scale stoichiometric model reconstructions using a simple web API. There are `108` models of intracellular biochemistry occurring in various organisms (including humans) in the database (so far); [see here for a list of models](http://bigg.ucsd.edu/models).\n",
    "> * Here, we'll first explore the [core metabolic model of Palsson and coworkers](https://pubmed.ncbi.nlm.nih.gov/26443778/), which is a scaled-down model of [carbohydrate metabolism](https://en.wikipedia.org/wiki/Carbohydrate_metabolism) in _E.coli_. This model has 72 metabolites and 95 reactions. We'll then look at other models, and see what is going on with these.\n",
    "> \n",
    ">  We call the model download endpoint of [the BiGG models API](http://bigg.ucsd.edu/data_access) and then save the model file to disk (so we don't hit the API unless we have to). \n",
    "\n",
    "This call returns model information organized as [a Julia dictionary](https://docs.julialang.org/en/v1/base/collections/#Base.Dict) in the `model::Dict{String, Any}` variable. If a model file is saved, we use the cached file instead of making an API call."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5651b4e0-a56f-43d7-b350-3caeb68bed66",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "JSON.Object{String, Any} with 6 entries:\n",
       "  \"metabolites\"  => Any[Object{String, Any}(\"id\"=>\"13dpg_c\", \"name\"=>\"3-Phospho…\n",
       "  \"reactions\"    => Any[Object{String, Any}(\"id\"=>\"SK_pchol_hs_18_1_18_1_c\", \"n…\n",
       "  \"genes\"        => Any[Object{String, Any}(\"id\"=>\"Nrk1_AT1\", \"name\"=>\"NMRK1\", …\n",
       "  \"id\"           => \"iAB_RBC_283\"\n",
       "  \"compartments\" => Object{String, Any}(\"c\"=>\"cytosol\", \"e\"=>\"extracellular spa…\n",
       "  \"version\"      => \"1\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "model = let\n",
    "\n",
    "    # build download endpoint -\n",
    "    baseurl = \"http://bigg.ucsd.edu\"; # base url to download model\n",
    "    modelid = \"iAB_RBC_283\"; # model id to download\n",
    "    path_to_saved_model_file = joinpath(_PATH_TO_DATA, \"saved-model-$(modelid).jld2\");\n",
    "\n",
    "    # check: do we have a model file saved?\n",
    "    model = nothing;\n",
    "    if (isfile(path_to_saved_model_file) == false)\n",
    "        \n",
    "        endpoint = MyBiggModelsDownloadModelEndpointModel();\n",
    "        endpoint.bigg_id = modelid;\n",
    "        url = build(baseurl, endpoint)\n",
    "        model = MyBiggModelsDownloadModelEndpointModel(url);\n",
    "\n",
    "        # Before we move on, save this model for later (so we don't keep hitting the API)\n",
    "        save(path_to_saved_model_file, Dict(\"model\" => model));\n",
    "    else\n",
    "        model = load(path_to_saved_model_file)[\"model\"];\n",
    "    end\n",
    "    model; # return the model (either saved, or downloaded)\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fddedbf8-8031-492f-89ee-cfce4a37facc",
   "metadata": {},
   "source": [
    "__Metabolite records__: Each metabolite (chemical compound) in the network has an associated metabolite record with several fields. Let's take a look at the metabolite at index `1`. The key field for today in the metabolite record is the `id` field, an abbreviation or symbol associated with this metabolite."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9afddf7b-38d8-4330-8334-5bdf7fdea790",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "JSON.Object{String, Any} with 7 entries:\n",
       "  \"id\"          => \"13dpg_c\"\n",
       "  \"name\"        => \"3-Phospho-D-glyceroyl phosphate\"\n",
       "  \"compartment\" => \"c\"\n",
       "  \"charge\"      => -4\n",
       "  \"formula\"     => \"C3H4O10P2\"\n",
       "  \"notes\"       => Object{String, Any}(\"original_bigg_ids\"=>Any[\"13dpg_c\"])\n",
       "  \"annotation\"  => Object{String, Any}(\"bigg.metabolite\"=>Any[\"13dpg\"], \"biocyc…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "model[\"metabolites\"][1] # example metabolite record"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "105f7c6c-af25-401e-b4c1-acc0bb22ed62",
   "metadata": {},
   "source": [
    "__Reaction records__: Similarly, each reaction in the network has a reaction record with several fields. Let's look at the reaction record at index `25`. The key field for the reaction record is the `metabolites` field, which lists the stoichiometric coefficients associated with this particular reaction."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9a972ff2-5203-4aca-ab7c-3fda49f8162c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "JSON.Object{String, Any} with 9 entries:\n",
       "  \"id\"                 => \"ACNAM9PL\"\n",
       "  \"name\"               => \"N-Acetylneuraminate 9-phosphate pyruvate-lyase (pyru…\n",
       "  \"metabolites\"        => Object{String, Any}(\"acmanap_c\"=>-1.0, \"acnamp_c\"=>1.…\n",
       "  \"lower_bound\"        => 0.0\n",
       "  \"upper_bound\"        => 1000.0\n",
       "  \"gene_reaction_rule\" => \"Nans_AT1\"\n",
       "  \"subsystem\"          => \"Aminosugar Metabolism\"\n",
       "  \"notes\"              => Object{String, Any}(\"original_bigg_ids\"=>Any[\"ACNAM9P…\n",
       "  \"annotation\"         => Object{String, Any}(\"bigg.reaction\"=>Any[\"ACNAM9PL\"],…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "model[\"reactions\"][25] # example reaction record"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d3a9029a-6650-42bb-a27a-a488fb424cee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "JSON.Object{String, Any} with 5 entries:\n",
       "  \"acmanap_c\" => -1.0\n",
       "  \"acnamp_c\"  => 1.0\n",
       "  \"h2o_c\"     => -1.0\n",
       "  \"pep_c\"     => -1.0\n",
       "  \"pi_c\"      => 1.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "model[\"reactions\"][25][\"metabolites\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d4c4971-c938-4bf5-b060-97f487a47812",
   "metadata": {},
   "source": [
    "__Stoichiometric matrix__: Next, let's build a stoichiometric matrix $\\mathbf{S}$ using the metabolite and reaction records. We'll do this using two for loops. \n",
    "\n",
    "> __Strategy__: In the outer loop, we iterate over the system's metabolites (chemical species) and select the `id` from the metabolites record for each metabolite. In the inner loop, we iterate over each reaction. For each reaction record, we ask if this reaction has an entry for the current metabolite `id` value; if it does, we grab the stoichiometric coefficient $\\sigma_{ij}$ corresponding to this metabolite and reaction.\n",
    "\n",
    "We'll save the stoichiometric matrix in the `S::Matrix{Float64}` variable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "59ccfc68-3498-4cac-bed6-846ad7281736",
   "metadata": {},
   "outputs": [],
   "source": [
    "S = let\n",
    "\n",
    "    # get some data from the model -\n",
    "    m = model[\"metabolites\"]; # get list of metabolites\n",
    "    r = model[\"reactions\"]; # get list of reactions\n",
    "    number_of_rows = length(m); # how many metabolites do we have? (rows)\n",
    "    number_of_cols = length(r); # how many reactions do we have? (cols)\n",
    "    S = zeros(number_of_rows,number_of_cols); # initialize an empty stoichiometric matrix\n",
    "\n",
    "    # let's build a stm -\n",
    "    for i ∈ eachindex(m)\n",
    "        metabolite = m[i][\"id\"]; # we are checking if this metabolite is in the reaction record\n",
    "        for j ∈ eachindex(r)\n",
    "            reaction = r[j];\n",
    "            if (haskey(reaction[\"metabolites\"], metabolite) == true)\n",
    "                S[i,j] = reaction[\"metabolites\"][metabolite];\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "    S; \n",
    "end;"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a16eef3",
   "metadata": {},
   "source": [
    "___"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78cb483b",
   "metadata": {},
   "source": [
    "## Task 1: Compute the covariance matrix\n",
    "In this task, we'll compute the covariance matrix $\\mathbf{\\Sigma}$ of the stoichiometric matrix $\\mathbf{S}$.\n",
    "\n",
    "> __Why are we doing this?__ The stoichiometric matrix is not square. Thus, we cannot directly compute its eigendecomposition. However, suppose we compute [the covariance matrix](https://en.wikipedia.org/wiki/Covariance_matrix) between the columns, i.e., between the reactions stoichiometric vectors. This gives us an idea about the relationship between the network's reactions $i$ and $j$.\n",
    "\n",
    "Let's start by computing the covariance matrix $\\mathbf{\\Sigma}$ using [the `cov(...)` function](https://docs.julialang.org/en/v1/stdlib/Statistics/#Statistics.cov) from the `Statistics.jl` package (this will be our ground truth). We'll save this covariance matrix in the `Σ::Matrix{Float64}` variable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "5ee14d17-cdd9-4e91-8194-9b19a48bc13a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "469×469 Matrix{Float64}:\n",
       "  0.00292398  -8.57471e-6  -8.57471e-6  …   8.57471e-6   8.57471e-6\n",
       " -8.57471e-6   0.00292398  -8.57471e-6      8.57471e-6   8.57471e-6\n",
       " -8.57471e-6  -8.57471e-6   0.00292398      8.57471e-6   8.57471e-6\n",
       " -8.57471e-6  -8.57471e-6  -8.57471e-6      8.57471e-6   8.57471e-6\n",
       " -8.57471e-6  -8.57471e-6  -8.57471e-6      8.57471e-6   8.57471e-6\n",
       " -8.57471e-6  -8.57471e-6  -8.57471e-6  …   8.57471e-6   8.57471e-6\n",
       " -8.57471e-6  -8.57471e-6  -8.57471e-6     -0.00292398   8.57471e-6\n",
       " -8.57471e-6  -8.57471e-6  -8.57471e-6      8.57471e-6  -0.00292398\n",
       " -8.57471e-6  -8.57471e-6  -8.57471e-6      8.57471e-6   8.57471e-6\n",
       " -8.57471e-6  -8.57471e-6  -8.57471e-6      8.57471e-6   8.57471e-6\n",
       "  ⋮                                     ⋱               \n",
       "  8.57471e-6   8.57471e-6   8.57471e-6  …   0.00292398   0.00292398\n",
       " -8.57471e-6  -8.57471e-6  -8.57471e-6     -0.00292398  -0.00292398\n",
       "  1.71494e-5   1.71494e-5   1.71494e-5      0.00584795   0.00584795\n",
       "  8.57471e-6  -0.00292398   8.57471e-6      0.00585653   0.00585653\n",
       "  8.57471e-6   8.57471e-6  -0.00292398      0.00585653   0.00585653\n",
       "  8.57471e-6   8.57471e-6   8.57471e-6  …   0.00585653   0.00585653\n",
       "  8.57471e-6   8.57471e-6   8.57471e-6      0.00878908   0.00878908\n",
       "  8.57471e-6   8.57471e-6   8.57471e-6      0.0146542    0.00878908\n",
       "  8.57471e-6   8.57471e-6   8.57471e-6      0.00878908   0.0146542"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "Σ = cov(S) # this is col covariance matrix from Statistics.jl: Ground truth!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3d36013-c215-402d-b1ae-b5b837198907",
   "metadata": {},
   "source": [
    "Next, let's compute the empirical covariance matrix $\\hat{\\mathbf{\\Sigma}}\\in\\mathbb{R}^{|\\mathcal{R}|\\times|\\mathcal{R}|}$. The empirical covariance matrix is computed as:\n",
    "$$\n",
    "\\hat{\\mathbf{\\Sigma}} = \\frac{1}{n-1}\\tilde{\\mathbf{X}}^{\\top}\\tilde{\\mathbf{X}}\n",
    "$$\n",
    "where $\\tilde{\\mathbf{X}}\\in\\mathbb{R}^{n\\times{|\\mathcal{R}|}}$ is the mean-centered stoichiometric matrix (i.e., each column has zero mean), and $n$ is the number of metabolites (rows) in the stoichiometric matrix. We'll verify our implementation by comparing our matrix with the one computed using [the `cov(...)` function](https://docs.julialang.org/en/v1/stdlib/Statistics/#Statistics.cov) from the `Statistics.jl` package.\n",
    "\n",
    "Let's save the covariance matrix in the `Σ̂::Matrix{Float64}` variable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "531b11a4-6234-49b7-875f-21b8be209733",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "469×469 Matrix{Float64}:\n",
       "  0.00292398  -8.57471e-6  -8.57471e-6  …   8.57471e-6   8.57471e-6\n",
       " -8.57471e-6   0.00292398  -8.57471e-6      8.57471e-6   8.57471e-6\n",
       " -8.57471e-6  -8.57471e-6   0.00292398      8.57471e-6   8.57471e-6\n",
       " -8.57471e-6  -8.57471e-6  -8.57471e-6      8.57471e-6   8.57471e-6\n",
       " -8.57471e-6  -8.57471e-6  -8.57471e-6      8.57471e-6   8.57471e-6\n",
       " -8.57471e-6  -8.57471e-6  -8.57471e-6  …   8.57471e-6   8.57471e-6\n",
       " -8.57471e-6  -8.57471e-6  -8.57471e-6     -0.00292398   8.57471e-6\n",
       " -8.57471e-6  -8.57471e-6  -8.57471e-6      8.57471e-6  -0.00292398\n",
       " -8.57471e-6  -8.57471e-6  -8.57471e-6      8.57471e-6   8.57471e-6\n",
       " -8.57471e-6  -8.57471e-6  -8.57471e-6      8.57471e-6   8.57471e-6\n",
       "  ⋮                                     ⋱               \n",
       "  8.57471e-6   8.57471e-6   8.57471e-6  …   0.00292398   0.00292398\n",
       " -8.57471e-6  -8.57471e-6  -8.57471e-6     -0.00292398  -0.00292398\n",
       "  1.71494e-5   1.71494e-5   1.71494e-5      0.00584795   0.00584795\n",
       "  8.57471e-6  -0.00292398   8.57471e-6      0.00585653   0.00585653\n",
       "  8.57471e-6   8.57471e-6  -0.00292398      0.00585653   0.00585653\n",
       "  8.57471e-6   8.57471e-6   8.57471e-6  …   0.00585653   0.00585653\n",
       "  8.57471e-6   8.57471e-6   8.57471e-6      0.00878908   0.00878908\n",
       "  8.57471e-6   8.57471e-6   8.57471e-6      0.0146542    0.00878908\n",
       "  8.57471e-6   8.57471e-6   8.57471e-6      0.00878908   0.0146542"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "Σ̂ = let\n",
    "\n",
    "    # initialize -\n",
    "    number_of_metabolites = size(S,1); # rows are metabolites\n",
    "    number_of_reactions = size(S,2); # cols are reactions\n",
    "    m = mean(S, dims=1) |> vec; # mean of each column\n",
    "    ones_vector = ones(number_of_metabolites); # ones vector\n",
    "    B = ones_vector * transpose(m); # outer product to center data\n",
    "\n",
    "    X̂ = S - B; # centered data matrix\n",
    "    Σ̂ = (transpose(X̂) * X̂) / (number_of_metabolites - 1); # compute covariance matrix\n",
    "    Σ̂; # return covariance matrix\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f08e63e8",
   "metadata": {},
   "source": [
    "### Check: Covariance matricies the same?\n",
    "Let's check that our computed covariance matrix `Σ̂` is the same as the one computed using [the `cov(...)` function](https://docs.julialang.org/en/v1/stdlib/Statistics/#Statistics.cov) from the `Statistics.jl` package. We'll use [the `isapprox(...)` function](https://docs.julialang.org/en/v1/base/math/#Base.isapprox) to check if the two matrices are approximately equal."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "f773f4c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "let\n",
    "    # initialize -\n",
    "    atol = 1e-8; # absolute tolerance for comparison\n",
    "    is_equal = isapprox(Σ, Σ̂; atol = atol);\n",
    "    @assert is_equal == true \"Covariance matrices are not approximately equal!\";\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "675f9265",
   "metadata": {},
   "source": [
    "Did we blow up? If not, great! Our covariance matrix implementation is correct. Let's move on and compute the eigendecomposition of the emphirical covariance matrix."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "524f7bca",
   "metadata": {},
   "source": [
    "___"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3d0585c",
   "metadata": {},
   "source": [
    "## Task 2: Compute the largest eigenvalue/eigenvector\n",
    "In this task, we'll compute the largest eigenvalue and associated eigenvector of the $\\mathbf{\\Sigma}$ matrix using the [power iteration method](https://en.wikipedia.org/wiki/Power_iteration). \n",
    "\n",
    "However, before we do that, let's use Julia's built-in [eigendecomposition function `eigen(...)`](https://docs.julialang.org/en/v1/stdlib/LinearAlgebra/#LinearAlgebra.eigen) to compute all the eigenvalues and eigenvectors of the $\\mathbf{\\Sigma}$ matrix (so we have something to compare against).\n",
    "\n",
    "> __Why compute all the values__? We'll use [the `eigen(...)` method](https://docs.julialang.org/en/v1/stdlib/LinearAlgebra/#LinearAlgebra.eigen) and then grab the last value(s) when we look at the largest value, this method returns all the eigenvalues and eigenvectors, sorted in ascending order. Thus, we only really need the last value for this example, but later we'll use the rest. \n",
    "\n",
    "We'll store the eigenvalues in the diagonal $\\mathbf{\\Lambda}$ matrix, while the eigenvectors will be stored in the $\\mathbf{V}$ matrix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "19bbb025-3688-4938-be93-b3a32358a837",
   "metadata": {},
   "outputs": [],
   "source": [
    "Λ,V = let\n",
    "\n",
    "    # initialize -\n",
    "    A = Σ̂; # this is the matrix that we will decompose (if we are looking Σ̂, then change)\n",
    "    (n,m) = size(A); # what is the dimension of A?\n",
    "    Λ = Matrix{Float64}(1.0*I, n, n); # builds the I matrix, we'll update with λ -\n",
    "    \n",
    "    # Decompose using the built-in function\n",
    "    F = eigen(A);   # eigenvalues and vectors in F of type Eigen\n",
    "    λ = F.values;   # vector of eigenvalues\n",
    "    V = F.vectors;  # n x n matrix of eigenvectors, each col is an eigenvector\n",
    "\n",
    "    # package the eigenvalues into Λ -\n",
    "    for i ∈ 1:n\n",
    "        Λ[i,i] = λ[i];\n",
    "    end\n",
    "\n",
    "    Λ,V\n",
    "end;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2bc657f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.1581270552146485"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "Λ[end,end] # largest eigenvalue"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3bbe74ab",
   "metadata": {},
   "source": [
    "Now let's compute the largest eigenvalue and eigenvector using the power iteration method.\n",
    "\n",
    "> __What is going on here?__ The power iteration method iteratively computes the dominant eigenpair by multiplying a guess vector by matrix $\\mathbf{A}$ and normalizing. We specify the matrix `A::Matrix{Float64}`, initial vector `v::Vector{Float64}`, maximum iterations `maxiter::Int`, and convergence tolerance `ϵ::Float64`.\n",
    "\n",
    "We'll save the largest eigenvalue in the `λ₁::Float64` variable and the associated eigenvector in the `v₁::Vector{Float64}` variable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "801dcd29",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converged in 11 iterations\n"
     ]
    }
   ],
   "source": [
    "λ₁,v₁ = let\n",
    "\n",
    "    # initialize -\n",
    "    A = Σ̂; # this is the matrix that we will decompose (free to change)\n",
    "    (n,m) = size(A); # what is the dimension of A?\n",
    "    v = ones(n); # initial guess for the eigenvector\n",
    "    maxiter = 1000; # maxinum number of iterations\n",
    "    ϵ = 0.000000001; # tolerance\n",
    "\n",
    "    # call poweriteration method\n",
    "    result = poweriteration(A,v, maxiter = maxiter, ϵ = ϵ);\n",
    "\n",
    "    # pull data from result \n",
    "    (result.value, result.vector)\n",
    "end;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "bf8d699c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.1581270548611515"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "λ₁ # largest eigenvalue estimate from power iteration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a6b3ce84",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "469-element Vector{Float64}:\n",
       " -0.0001379918384368964\n",
       " -0.00013759551475015237\n",
       " -0.00013759551477686842\n",
       " -0.00013799284710265552\n",
       "  1.6149740442652996e-5\n",
       "  1.6149737852061186e-5\n",
       "  1.614974043608122e-5\n",
       "  1.6149740442957386e-5\n",
       "  1.6149743026790564e-5\n",
       "  1.6149743026790405e-5\n",
       "  ⋮\n",
       " -0.0417886037087952\n",
       "  0.04486310194343365\n",
       " -0.07461340171986786\n",
       " -0.040978740363656806\n",
       " -0.040978740363630084\n",
       " -0.04097834201340546\n",
       " -0.04113527318425875\n",
       " -0.041135274202159566\n",
       " -0.041135274204795576"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "V[:,end] # largest eigenvector from built-in function"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b85b508a",
   "metadata": {},
   "source": [
    "To test our power iteration implementation, let's compare the values of the largest eigenvalue, eigenvector pair $(\\lambda_{1}, \\mathbf{v}_{1})$ that we just estimated with those computed [using the `eigen(...)` function exported by the LinearAlgebra.jl package](https://docs.julialang.org/en/v1/stdlib/LinearAlgebra/#LinearAlgebra.eigen). \n",
    "\n",
    "> __Check__: We use the [`@assert` macro](https://docs.julialang.org/en/v1/base/base/#Base.@assert) in combination with the shortcut version of [the `isapprox(...)` function](https://docs.julialang.org/en/v1/base/math/#Base.isapprox) to compare our result to the built in function. If the argument to [the `@assert` macro](https://docs.julialang.org/en/v1/base/base/#Base.@assert) evaluates to `false`, an [`AssertionError` instance](https://docs.julialang.org/en/v1/base/base/#Core.AssertionError) is thrown. We'll compare the values to within an absolute tolerance value of `atol::Float64`.\n",
    "\n",
    "So does our power iteration method work?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "fcf00e15",
   "metadata": {},
   "outputs": [],
   "source": [
    "let\n",
    "    \n",
    "    # initialize -\n",
    "    λ = Λ[end,end]; # largest eigenvalue is in the last row/col of Λ\n",
    "    v = V[:,end]; # largest eigenvector is in the last column of V\n",
    "    atol = 1e-4; # tolerance for comparison\n",
    "\n",
    "    # tests\n",
    "    @assert isapprox(λ, λ₁, atol = atol)  # do the eigenvalues and eigenvectors match?\n",
    "    @assert isapprox(abs.(v), abs.(v₁), atol = atol); # do the eigenvectors match?\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31ea930f",
   "metadata": {},
   "source": [
    "If we get here, without an error (and a reasonable tolerance), our power iteration method works! Now that we have confidence in our implementation, let's explore what the largest eigenvalue/eigenvector pair tells us about the stoichiometric matrix $\\mathbf{S}$ and the reaction network it represents.\n",
    "\n",
    "___"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74a30aeb",
   "metadata": {},
   "source": [
    "## Task 3: What does the largest eigenvalue/eigenvector tell us?\n",
    "In this task, we'll explore what the largest eigenvector coefficients tell us about the $\\mathbf{\\Sigma}$ matrix and, by extension, the stoichiometric matrix $\\mathbf{S}$. Is there something interesting here? First, let's grab the absolute largest component, and see what reaction it corresponds to.\n",
    "> __Largest absolute component__ \n",
    "> \n",
    "> The largest component of the largest eigenvector $\\mathbf{v}_{1}$ tells us which reaction is (loosely) the most important in the network. However, the components of the eigenvector can be positive or negative, so we need to be a little careful here, and just look at the absolute values of the components. \n",
    "> \n",
    "> For some eigenvector $\\mathbf{z}$, the ith scaled component $\\sigma(\\mathbf{z})_{i}$ is given by:\n",
    "> $$\n",
    "\\begin{align*}\n",
    "    \\sigma(\\mathbf{z})_{i} &= \\frac{\\text{abs}(z_{i})}{\\sum_{j=1}^{m}\\text{abs}(z_{j})}\\quad{i=1,2,\\dots,m}\n",
    "\\end{align*}    \n",
    "$$\n",
    "> The scaled vectors should sum to `1`; thus, we can think about the elements (loosely) as probabilities, i.e., the probability that the ith component is the most important. \n",
    "\n",
    "Let's do the scaling, grab the index of the maximum scaled element [using the `argmax(...)` method](https://docs.julialang.org/en/v1/base/collections/#Base.argmax). We'll then pipe `|>` that index into reactions data, and pull out the most important reaction record:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "605ca446",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "JSON.Object{String, Any} with 9 entries:\n",
       "  \"id\"                 => \"HOXG\"\n",
       "  \"name\"               => \"Heme oxygenase 1\"\n",
       "  \"metabolites\"        => Object{String, Any}(\"biliverd_c\"=>1.0, \"co_c\"=>1.0, \"…\n",
       "  \"lower_bound\"        => 0.0\n",
       "  \"upper_bound\"        => 1000.0\n",
       "  \"gene_reaction_rule\" => \"Hmox2_AT1 or Hmox1_AT1\"\n",
       "  \"subsystem\"          => \"Heme Degradation\"\n",
       "  \"notes\"              => Object{String, Any}(\"original_bigg_ids\"=>Any[\"HOXG\"])\n",
       "  \"annotation\"         => Object{String, Any}(\"bigg.reaction\"=>Any[\"HOXG\"], \"ec…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sum(abs.(v₁)) |> T -> (1/T)*abs.(v₁) |> argmax |> i-> model[\"reactions\"][i]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "929b969a",
   "metadata": {},
   "source": [
    "Next, let's look at the top `k` reactions in the network. The following code identifies the reactions with the largest absolute eigenvector components after normalization. By sorting the normalized values, we can identify which reactions contribute most to the dominant eigenvector, providing insights into the network's structure.\n",
    "\n",
    "> __What is going on here?__ We normalize the absolute eigenvector values to create importance weights, then use [sortperm(...)](https://docs.julialang.org/en/v1/base/sort/#Base.sortperm) to identify the top `k::Int64` reactions ranked by contribution.\n",
    "\n",
    "Let's save the top `k::Int64` reactions in the `top_k_reactions::Vector{JSON.Object{String, Any}}` array."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "aa34c731",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cumulative contribution of top 10 reactions: 12.02%\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "10-element Vector{JSON.Object{String, Any}}:\n",
       " JSON.Object(\"id\" => \"HOXG\", \"name\" => \"Heme oxygenase 1\", \"metabolites\" => JSON.Object{String, Any}(\"biliverd_c\" => 1.0, \"co_c\" => 1.0, \"fe2_c\" => 1.0, \"h2o_c\" => 3.0, \"h_c\" => -5.0, \"nadp_c\" => 3.0, \"nadph_c\" => -3.0, \"o2_c\" => -3.0, \"pheme_c\" => -1.0), \"lower_bound\" => 0.0, \"upper_bound\" => 1000.0, \"gene_reaction_rule\" => \"Hmox2_AT1 or Hmox1_AT1\", \"subsystem\" => \"Heme Degradation\", \"notes\" => JSON.Object{String, Any}(\"original_bigg_ids\" => Any[\"HOXG\"]), \"annotation\" => JSON.Object{String, Any}(\"bigg.reaction\" => Any[\"HOXG\"], \"ec-code\" => Any[\"1.14.99.3\"], \"metanetx.reaction\" => Any[\"MNXR100681\"], \"sbo\" => \"SBO:0000176\"))\n",
       " JSON.Object(\"id\" => \"3MOXTYRESSte\", \"name\" => \"3-Methoxytyramine secretion via secretory vesicle (ATP driven)\", \"metabolites\" => JSON.Object{String, Any}(\"3moxtyr_c\" => -3.0, \"3moxtyr_e\" => 3.0, \"adp_c\" => 2.0, \"atp_c\" => -2.0, \"h2o_c\" => -2.0, \"h_c\" => 2.0, \"pi_c\" => 2.0), \"lower_bound\" => 0.0, \"upper_bound\" => 1000.0, \"gene_reaction_rule\" => \"\", \"subsystem\" => \"Transport, Extracellular\", \"notes\" => JSON.Object{String, Any}(\"original_bigg_ids\" => Any[\"3MOXTYRESSte\"]), \"annotation\" => JSON.Object{String, Any}(\"bigg.reaction\" => Any[\"3MOXTYRESSte\"], \"metanetx.reaction\" => Any[\"MNXR94930\"], \"sbo\" => \"SBO:0000185\"))\n",
       " JSON.Object(\"id\" => \"NORMETEVESSte\", \"name\" => \"L-Normetanephrine seceretion via secretory vesicle (ATP driven)\", \"metabolites\" => JSON.Object{String, Any}(\"adp_c\" => 2.0, \"atp_c\" => -2.0, \"h2o_c\" => -2.0, \"h_c\" => 2.0, \"normete__L_c\" => -3.0, \"normete__L_e\" => 3.0, \"pi_c\" => 2.0), \"lower_bound\" => 0.0, \"upper_bound\" => 1000.0, \"gene_reaction_rule\" => \"\", \"subsystem\" => \"Transport, Extracellular\", \"notes\" => JSON.Object{String, Any}(\"original_bigg_ids\" => Any[\"NORMETEVESSte\"]), \"annotation\" => JSON.Object{String, Any}(\"bigg.reaction\" => Any[\"NORMETEVESSte\"], \"metanetx.reaction\" => Any[\"MNXR102002\"], \"sbo\" => \"SBO:0000185\"))\n",
       " JSON.Object(\"id\" => \"MEPIVESSte\", \"name\" => \"Metanephrine secretion via secretory vesicle (ATP driven)\", \"metabolites\" => JSON.Object{String, Any}(\"adp_c\" => 2.0, \"atp_c\" => -2.0, \"h2o_c\" => -2.0, \"h_c\" => 2.0, \"mepi_c\" => -3.0, \"mepi_e\" => 3.0, \"pi_c\" => 2.0), \"lower_bound\" => 0.0, \"upper_bound\" => 1000.0, \"gene_reaction_rule\" => \"\", \"subsystem\" => \"Transport, Extracellular\", \"notes\" => JSON.Object{String, Any}(\"original_bigg_ids\" => Any[\"MEPIVESSte\"]), \"annotation\" => JSON.Object{String, Any}(\"bigg.reaction\" => Any[\"MEPIVESSte\"], \"metanetx.reaction\" => Any[\"MNXR101465\"], \"sbo\" => \"SBO:0000185\"))\n",
       " JSON.Object(\"id\" => \"UPPDC1\", \"name\" => \"Uroporphyrinogen decarboxylase (uroporphyrinogen III)\", \"metabolites\" => JSON.Object{String, Any}(\"co2_c\" => 4.0, \"cpppg3_c\" => 1.0, \"h_c\" => -4.0, \"uppg3_c\" => -1.0), \"lower_bound\" => 0.0, \"upper_bound\" => 1000.0, \"gene_reaction_rule\" => \"Urod_AT1\", \"subsystem\" => \"Heme Biosynthesis\", \"notes\" => JSON.Object{String, Any}(\"original_bigg_ids\" => Any[\"UPPDC1\"]), \"annotation\" => JSON.Object{String, Any}(\"bigg.reaction\" => Any[\"UPPDC1\"], \"biocyc\" => Any[\"META:UROGENDECARBOX-RXN\"], \"ec-code\" => Any[\"4.1.1.37\"], \"kegg.reaction\" => Any[\"R03197\"], \"metanetx.reaction\" => Any[\"MNXR105140\"], \"reactome.reaction\" => Any[\"R-GGA-421515\", \"R-DME-189425\", \"R-BTA-189425\", \"R-SSC-189425\", \"R-DDI-189425\", \"R-RNO-189425\", \"R-SPO-189425\", \"R-OSA-189425\", \"R-GGA-189425\", \"R-HSA-189425\", \"R-SCE-189425\", \"R-MMU-189425\", \"R-PFA-189425\", \"R-XTR-189425\", \"R-TGU-189425\", \"R-CFA-189425\", \"R-DRE-189425\", \"R-ATH-189425\"], \"rhea\" => Any[\"19868\", \"19865\", \"19867\", \"19866\"], \"sbo\" => \"SBO:0000176\", \"seed.reaction\" => Any[\"rxn02288\"]))\n",
       " JSON.Object(\"id\" => \"UDPGD\", \"name\" => \"UDPglucose 6-dehydrogenase\", \"metabolites\" => JSON.Object{String, Any}(\"h2o_c\" => -1.0, \"h_c\" => 3.0, \"nad_c\" => -2.0, \"nadh_c\" => 2.0, \"udpg_c\" => -1.0, \"udpglcur_c\" => 1.0), \"lower_bound\" => 0.0, \"upper_bound\" => 1000.0, \"gene_reaction_rule\" => \"Ugdh_AT1\", \"subsystem\" => \"Starch and Sucrose Metabolism\", \"notes\" => JSON.Object{String, Any}(\"original_bigg_ids\" => Any[\"UDPGD\"]), \"annotation\" => JSON.Object{String, Any}(\"bigg.reaction\" => Any[\"UDPGD\"], \"biocyc\" => Any[\"META:UGD-RXN\"], \"ec-code\" => Any[\"1.1.1.22\"], \"kegg.reaction\" => Any[\"R00286\"], \"metanetx.reaction\" => Any[\"MNXR141783\"], \"reactome.reaction\" => Any[\"R-TGU-173597\", \"R-SSC-173597\", \"R-OSA-173597\", \"R-RNO-173597\", \"R-GGA-173597\", \"R-MMU-173597\", \"R-XTR-173597\", \"R-DRE-173597\", \"R-CEL-173597\", \"R-CFA-173597\", \"R-BTA-173597\", \"R-DME-173597\", \"R-ATH-173597\", \"R-HSA-173597\"], \"rhea\" => Any[\"23598\", \"23599\", \"23597\", \"23596\"], \"sabiork\" => Any[\"8582\", \"784\"], \"sbo\" => \"SBO:0000176\", \"seed.reaction\" => Any[\"rxn00211\"]…))\n",
       " JSON.Object(\"id\" => \"CPPPGO\", \"name\" => \"Coproporphyrinogen oxidase (O2 required)\", \"metabolites\" => JSON.Object{String, Any}(\"co2_c\" => 2.0, \"cpppg3_c\" => -1.0, \"h2o_c\" => 2.0, \"h_c\" => -2.0, \"o2_c\" => -1.0, \"pppg9_c\" => 1.0), \"lower_bound\" => 0.0, \"upper_bound\" => 1000.0, \"gene_reaction_rule\" => \"Cpox_AT1\", \"subsystem\" => \"Heme Biosynthesis\", \"notes\" => JSON.Object{String, Any}(\"original_bigg_ids\" => Any[\"CPPPGO\"]), \"annotation\" => JSON.Object{String, Any}(\"bigg.reaction\" => Any[\"CPPPGO\"], \"biocyc\" => Any[\"META:RXN0-1461\"], \"ec-code\" => Any[\"1.3.3.3\"], \"kegg.reaction\" => Any[\"R03220\"], \"metanetx.reaction\" => Any[\"MNXR96880\"], \"rhea\" => Any[\"18257\", \"18258\", \"18259\", \"18260\"], \"sbo\" => \"SBO:0000176\", \"seed.reaction\" => Any[\"rxn02303\"]))\n",
       " JSON.Object(\"id\" => \"OPAH\", \"name\" => \"5-Oxoproline amidohydrolase (ATP-hydrolysing) (ir)\", \"metabolites\" => JSON.Object{String, Any}(\"5oxpro_c\" => -1.0, \"adp_c\" => 1.0, \"atp_c\" => -1.0, \"glu__L_c\" => 1.0, \"h2o_c\" => -2.0, \"h_c\" => 1.0, \"pi_c\" => 1.0), \"lower_bound\" => 0.0, \"upper_bound\" => 1000.0, \"gene_reaction_rule\" => \"Oplah_AT1\", \"subsystem\" => \"Glutathione Metabolism\", \"notes\" => JSON.Object{String, Any}(\"original_bigg_ids\" => Any[\"OPAHir\"]), \"annotation\" => JSON.Object{String, Any}(\"bigg.reaction\" => Any[\"OPAH\"], \"biocyc\" => Any[\"META:5-OXOPROLINASE-ATP-HYDROLYSING-RXN\"], \"ec-code\" => Any[\"3.5.2.9\"], \"kegg.reaction\" => Any[\"R00251\"], \"metanetx.reaction\" => Any[\"MNXR138674\"], \"reactome.reaction\" => Any[\"R-SPO-1247935\", \"R-OSA-1247935\", \"R-HSA-1247935\", \"R-ATH-1247935\", \"R-DDI-1247935\", \"R-BTA-1247935\", \"R-DME-1247935\", \"R-CFA-1247935\", \"R-DRE-1247935\", \"R-SCE-1247935\", \"R-XTR-1247935\", \"R-RNO-1247935\", \"R-CEL-1247935\", \"R-SSC-1247935\", \"R-MMU-1247935\"], \"rhea\" => Any[\"10348\", \"10349\", \"10350\", \"10351\"], \"sabiork\" => Any[\"8623\"], \"sbo\" => \"SBO:0000176\", \"seed.reaction\" => Any[\"rxn00186\"]…))\n",
       " JSON.Object(\"id\" => \"GMPS2\", \"name\" => \"GMP synthase\", \"metabolites\" => JSON.Object{String, Any}(\"amp_c\" => 1.0, \"atp_c\" => -1.0, \"gln__L_c\" => -1.0, \"glu__L_c\" => 1.0, \"gmp_c\" => 1.0, \"h2o_c\" => -1.0, \"h_c\" => 2.0, \"ppi_c\" => 1.0, \"xmp_c\" => -1.0), \"lower_bound\" => 0.0, \"upper_bound\" => 1000.0, \"gene_reaction_rule\" => \"Gmps_AT1\", \"subsystem\" => \"Nucleotides\", \"notes\" => JSON.Object{String, Any}(\"original_bigg_ids\" => Any[\"GMPS2\"]), \"annotation\" => JSON.Object{String, Any}(\"bigg.reaction\" => Any[\"GMPS2\"], \"biocyc\" => Any[\"META:GMP-SYN-GLUT-RXN\"], \"ec-code\" => Any[\"6.3.5.2\"], \"kegg.reaction\" => Any[\"R01231\"], \"metanetx.reaction\" => Any[\"MNXR100384\"], \"rhea\" => Any[\"11680\", \"11682\", \"11681\", \"11683\"], \"sabiork\" => Any[\"215\"], \"sbo\" => \"SBO:0000176\", \"seed.reaction\" => Any[\"rxn00917\"]))\n",
       " JSON.Object(\"id\" => \"NaKt\", \"name\" => \"Na+/K+ exchanging ATPase\", \"metabolites\" => JSON.Object{String, Any}(\"adp_c\" => 1.0, \"atp_c\" => -1.0, \"h2o_c\" => -1.0, \"h_c\" => 1.0, \"k_c\" => 2.0, \"k_e\" => -2.0, \"na1_c\" => -3.0, \"na1_e\" => 3.0, \"pi_c\" => 1.0), \"lower_bound\" => 0.0, \"upper_bound\" => 1000.0, \"gene_reaction_rule\" => \"Atp1a1b1 or Atp1a1b2 or Atp1a1b3 or Atp1a1b4\", \"objective_coefficient\" => 1.0, \"subsystem\" => \"Transport, Extracellular\", \"notes\" => JSON.Object{String, Any}(\"original_bigg_ids\" => Any[\"NaKt\"]), \"annotation\" => JSON.Object{String, Any}(\"bigg.reaction\" => Any[\"NaKt\"], \"biocyc\" => Any[\"META:3.6.3.9-RXN\"], \"ec-code\" => Any[\"3.6.3.9\"], \"metanetx.reaction\" => Any[\"MNXR102073\"], \"sbo\" => \"SBO:0000185\")…)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "top_k_reactions = let\n",
    "\n",
    "    # initialize -\n",
    "    k = 10; # how many top reactions do we want?\n",
    "    normalized_v₁ = sum(abs.(v₁)) |> T -> (1/T)*abs.(v₁); # normalize the eigenvector\n",
    "    sorted_indices = sortperm(normalized_v₁, rev = true); # get sorted indices (largest to smallest)\n",
    "    top_k_indices = sorted_indices[1:k]; # get top k indices\n",
    "    top_k_reactions = [model[\"reactions\"][i] for i ∈ top_k_indices]; # get top k reactions\n",
    "\n",
    "    # what is the cumulative contribution of these top k reactions?\n",
    "    cumulative_contribution = sum(normalized_v₁[top_k_indices]);\n",
    "    println(\"Cumulative contribution of top $k reactions: $(round(cumulative_contribution*100, digits=2))%\");\n",
    "\n",
    "    top_k_reactions # return top k reactions\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf7f7a46",
   "metadata": {},
   "source": [
    "Alternatively, let's use [the `softmax(...)` function](src/Compute.jl) to transform the largest eigenvector in a probability vector (sums to one, all entries are non-negative). The [softmax](https://en.wikipedia.org/wiki/Softmax_function) for some vector $\\mathbf{z}$ as\n",
    "$$\n",
    "\\begin{equation}\n",
    "\\sigma(\\mathbf{z})_{i} = \\frac{e^{z_{i}}}{\\sum_{j=1}^{m}e^{z_{j}}}\\quad{i=1,2,\\dots,m}\n",
    "\\end{equation}\n",
    "$$\n",
    "where $\\sigma(\\mathbf{z})_{i}$ is the ith components of the transformed eigenvector. We apply [the `argmax(...)` function](https://docs.julialang.org/en/v1/base/collections/#Base.argmax) to the transformed vector to get the largest single component, or we could grab the top `k` components as before."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f68d3bf5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cumulative contribution of top 10 reactions (softmax): 2.41%\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "10-element Vector{JSON.Object{String, Any}}:\n",
       " JSON.Object(\"id\" => \"3MOXTYRESSte\", \"name\" => \"3-Methoxytyramine secretion via secretory vesicle (ATP driven)\", \"metabolites\" => JSON.Object{String, Any}(\"3moxtyr_c\" => -3.0, \"3moxtyr_e\" => 3.0, \"adp_c\" => 2.0, \"atp_c\" => -2.0, \"h2o_c\" => -2.0, \"h_c\" => 2.0, \"pi_c\" => 2.0), \"lower_bound\" => 0.0, \"upper_bound\" => 1000.0, \"gene_reaction_rule\" => \"\", \"subsystem\" => \"Transport, Extracellular\", \"notes\" => JSON.Object{String, Any}(\"original_bigg_ids\" => Any[\"3MOXTYRESSte\"]), \"annotation\" => JSON.Object{String, Any}(\"bigg.reaction\" => Any[\"3MOXTYRESSte\"], \"metanetx.reaction\" => Any[\"MNXR94930\"], \"sbo\" => \"SBO:0000185\"))\n",
       " JSON.Object(\"id\" => \"NORMETEVESSte\", \"name\" => \"L-Normetanephrine seceretion via secretory vesicle (ATP driven)\", \"metabolites\" => JSON.Object{String, Any}(\"adp_c\" => 2.0, \"atp_c\" => -2.0, \"h2o_c\" => -2.0, \"h_c\" => 2.0, \"normete__L_c\" => -3.0, \"normete__L_e\" => 3.0, \"pi_c\" => 2.0), \"lower_bound\" => 0.0, \"upper_bound\" => 1000.0, \"gene_reaction_rule\" => \"\", \"subsystem\" => \"Transport, Extracellular\", \"notes\" => JSON.Object{String, Any}(\"original_bigg_ids\" => Any[\"NORMETEVESSte\"]), \"annotation\" => JSON.Object{String, Any}(\"bigg.reaction\" => Any[\"NORMETEVESSte\"], \"metanetx.reaction\" => Any[\"MNXR102002\"], \"sbo\" => \"SBO:0000185\"))\n",
       " JSON.Object(\"id\" => \"MEPIVESSte\", \"name\" => \"Metanephrine secretion via secretory vesicle (ATP driven)\", \"metabolites\" => JSON.Object{String, Any}(\"adp_c\" => 2.0, \"atp_c\" => -2.0, \"h2o_c\" => -2.0, \"h_c\" => 2.0, \"mepi_c\" => -3.0, \"mepi_e\" => 3.0, \"pi_c\" => 2.0), \"lower_bound\" => 0.0, \"upper_bound\" => 1000.0, \"gene_reaction_rule\" => \"\", \"subsystem\" => \"Transport, Extracellular\", \"notes\" => JSON.Object{String, Any}(\"original_bigg_ids\" => Any[\"MEPIVESSte\"]), \"annotation\" => JSON.Object{String, Any}(\"bigg.reaction\" => Any[\"MEPIVESSte\"], \"metanetx.reaction\" => Any[\"MNXR101465\"], \"sbo\" => \"SBO:0000185\"))\n",
       " JSON.Object(\"id\" => \"UDPGD\", \"name\" => \"UDPglucose 6-dehydrogenase\", \"metabolites\" => JSON.Object{String, Any}(\"h2o_c\" => -1.0, \"h_c\" => 3.0, \"nad_c\" => -2.0, \"nadh_c\" => 2.0, \"udpg_c\" => -1.0, \"udpglcur_c\" => 1.0), \"lower_bound\" => 0.0, \"upper_bound\" => 1000.0, \"gene_reaction_rule\" => \"Ugdh_AT1\", \"subsystem\" => \"Starch and Sucrose Metabolism\", \"notes\" => JSON.Object{String, Any}(\"original_bigg_ids\" => Any[\"UDPGD\"]), \"annotation\" => JSON.Object{String, Any}(\"bigg.reaction\" => Any[\"UDPGD\"], \"biocyc\" => Any[\"META:UGD-RXN\"], \"ec-code\" => Any[\"1.1.1.22\"], \"kegg.reaction\" => Any[\"R00286\"], \"metanetx.reaction\" => Any[\"MNXR141783\"], \"reactome.reaction\" => Any[\"R-TGU-173597\", \"R-SSC-173597\", \"R-OSA-173597\", \"R-RNO-173597\", \"R-GGA-173597\", \"R-MMU-173597\", \"R-XTR-173597\", \"R-DRE-173597\", \"R-CEL-173597\", \"R-CFA-173597\", \"R-BTA-173597\", \"R-DME-173597\", \"R-ATH-173597\", \"R-HSA-173597\"], \"rhea\" => Any[\"23598\", \"23599\", \"23597\", \"23596\"], \"sabiork\" => Any[\"8582\", \"784\"], \"sbo\" => \"SBO:0000176\", \"seed.reaction\" => Any[\"rxn00211\"]…))\n",
       " JSON.Object(\"id\" => \"OPAH\", \"name\" => \"5-Oxoproline amidohydrolase (ATP-hydrolysing) (ir)\", \"metabolites\" => JSON.Object{String, Any}(\"5oxpro_c\" => -1.0, \"adp_c\" => 1.0, \"atp_c\" => -1.0, \"glu__L_c\" => 1.0, \"h2o_c\" => -2.0, \"h_c\" => 1.0, \"pi_c\" => 1.0), \"lower_bound\" => 0.0, \"upper_bound\" => 1000.0, \"gene_reaction_rule\" => \"Oplah_AT1\", \"subsystem\" => \"Glutathione Metabolism\", \"notes\" => JSON.Object{String, Any}(\"original_bigg_ids\" => Any[\"OPAHir\"]), \"annotation\" => JSON.Object{String, Any}(\"bigg.reaction\" => Any[\"OPAH\"], \"biocyc\" => Any[\"META:5-OXOPROLINASE-ATP-HYDROLYSING-RXN\"], \"ec-code\" => Any[\"3.5.2.9\"], \"kegg.reaction\" => Any[\"R00251\"], \"metanetx.reaction\" => Any[\"MNXR138674\"], \"reactome.reaction\" => Any[\"R-SPO-1247935\", \"R-OSA-1247935\", \"R-HSA-1247935\", \"R-ATH-1247935\", \"R-DDI-1247935\", \"R-BTA-1247935\", \"R-DME-1247935\", \"R-CFA-1247935\", \"R-DRE-1247935\", \"R-SCE-1247935\", \"R-XTR-1247935\", \"R-RNO-1247935\", \"R-CEL-1247935\", \"R-SSC-1247935\", \"R-MMU-1247935\"], \"rhea\" => Any[\"10348\", \"10349\", \"10350\", \"10351\"], \"sabiork\" => Any[\"8623\"], \"sbo\" => \"SBO:0000176\", \"seed.reaction\" => Any[\"rxn00186\"]…))\n",
       " JSON.Object(\"id\" => \"GMPS2\", \"name\" => \"GMP synthase\", \"metabolites\" => JSON.Object{String, Any}(\"amp_c\" => 1.0, \"atp_c\" => -1.0, \"gln__L_c\" => -1.0, \"glu__L_c\" => 1.0, \"gmp_c\" => 1.0, \"h2o_c\" => -1.0, \"h_c\" => 2.0, \"ppi_c\" => 1.0, \"xmp_c\" => -1.0), \"lower_bound\" => 0.0, \"upper_bound\" => 1000.0, \"gene_reaction_rule\" => \"Gmps_AT1\", \"subsystem\" => \"Nucleotides\", \"notes\" => JSON.Object{String, Any}(\"original_bigg_ids\" => Any[\"GMPS2\"]), \"annotation\" => JSON.Object{String, Any}(\"bigg.reaction\" => Any[\"GMPS2\"], \"biocyc\" => Any[\"META:GMP-SYN-GLUT-RXN\"], \"ec-code\" => Any[\"6.3.5.2\"], \"kegg.reaction\" => Any[\"R01231\"], \"metanetx.reaction\" => Any[\"MNXR100384\"], \"rhea\" => Any[\"11680\", \"11682\", \"11681\", \"11683\"], \"sabiork\" => Any[\"215\"], \"sbo\" => \"SBO:0000176\", \"seed.reaction\" => Any[\"rxn00917\"]))\n",
       " JSON.Object(\"id\" => \"NaKt\", \"name\" => \"Na+/K+ exchanging ATPase\", \"metabolites\" => JSON.Object{String, Any}(\"adp_c\" => 1.0, \"atp_c\" => -1.0, \"h2o_c\" => -1.0, \"h_c\" => 1.0, \"k_c\" => 2.0, \"k_e\" => -2.0, \"na1_c\" => -3.0, \"na1_e\" => 3.0, \"pi_c\" => 1.0), \"lower_bound\" => 0.0, \"upper_bound\" => 1000.0, \"gene_reaction_rule\" => \"Atp1a1b1 or Atp1a1b2 or Atp1a1b3 or Atp1a1b4\", \"objective_coefficient\" => 1.0, \"subsystem\" => \"Transport, Extracellular\", \"notes\" => JSON.Object{String, Any}(\"original_bigg_ids\" => Any[\"NaKt\"]), \"annotation\" => JSON.Object{String, Any}(\"bigg.reaction\" => Any[\"NaKt\"], \"biocyc\" => Any[\"META:3.6.3.9-RXN\"], \"ec-code\" => Any[\"3.6.3.9\"], \"metanetx.reaction\" => Any[\"MNXR102073\"], \"sbo\" => \"SBO:0000185\")…)\n",
       " JSON.Object(\"id\" => \"ALDD2x\", \"name\" => \"Aldehyde dehydrogenase (acetaldehyde, NAD)\", \"metabolites\" => JSON.Object{String, Any}(\"ac_c\" => 1.0, \"acald_c\" => -1.0, \"h2o_c\" => -1.0, \"h_c\" => 2.0, \"nad_c\" => -1.0, \"nadh_c\" => 1.0), \"lower_bound\" => 0.0, \"upper_bound\" => 1000.0, \"gene_reaction_rule\" => \"Aldh1a2_AT2 or Aldh3a1_AT1 or Aldh9a1_AT1 or Aldh1a1_AT1\", \"subsystem\" => \"Glycolysis/Gluconeogenesis\", \"notes\" => JSON.Object{String, Any}(\"original_bigg_ids\" => Any[\"ALDD2x\"]), \"annotation\" => JSON.Object{String, Any}(\"bigg.reaction\" => Any[\"ALDD2x\"], \"biocyc\" => Any[\"META:RXN66-3\"], \"ec-code\" => Any[\"1.2.1.3\", \"1.2.1.5\"], \"kegg.reaction\" => Any[\"R00710\"], \"metanetx.reaction\" => Any[\"MNXR95749\"], \"reactome.reaction\" => Any[\"R-HSA-71723\", \"R-XTR-71691\", \"R-DRE-71723\", \"R-CEL-71691\", \"R-MMU-71723\", \"R-DME-71723\", \"R-SSC-5696091\", \"R-RNO-71691\", \"R-BTA-71723\", \"R-MMU-5696091\"  …  \"R-RNO-71723\", \"R-XTR-71723\", \"R-SCE-71723\", \"R-CFA-71723\", \"R-BTA-71691\", \"R-CFA-5696091\", \"R-CEL-71723\", \"R-SSC-71723\", \"R-GGA-71723\", \"R-GGA-71691\"], \"rhea\" => Any[\"25294\", \"25297\", \"25296\", \"25295\"], \"sabiork\" => Any[\"544\"], \"sbo\" => \"SBO:0000176\", \"seed.reaction\" => Any[\"rxn00506\"]…))\n",
       " JSON.Object(\"id\" => \"FADDP\", \"name\" => \"FAD diphosphatase\", \"metabolites\" => JSON.Object{String, Any}(\"amp_c\" => 1.0, \"fad_c\" => -1.0, \"fmn_c\" => 1.0, \"h2o_c\" => -1.0, \"h_c\" => 2.0), \"lower_bound\" => 0.0, \"upper_bound\" => 1000.0, \"gene_reaction_rule\" => \"EnpP1_AT1\", \"subsystem\" => \"Riboflavin Metabolism\", \"notes\" => JSON.Object{String, Any}(\"original_bigg_ids\" => Any[\"FADDP\"]), \"annotation\" => JSON.Object{String, Any}(\"bigg.reaction\" => Any[\"FADDP\"], \"biocyc\" => Any[\"META:FAD-PYROPHOSPHATASE-RXN\"], \"ec-code\" => Any[\"3.6.1.18\", \"3.6.1.9\"], \"kegg.reaction\" => Any[\"R00160\"], \"metanetx.reaction\" => Any[\"MNXR99209\"], \"reactome.reaction\" => Any[\"R-GGA-196955\", \"R-MMU-196955\", \"R-TGU-196955\", \"R-ATH-196955\", \"R-CFA-196955\", \"R-XTR-196955\", \"R-BTA-196955\", \"R-SSC-196955\", \"R-DRE-196955\", \"R-RNO-196955\", \"R-SPO-196955\", \"R-OSA-196955\", \"R-CEL-196955\", \"R-HSA-196955\", \"R-SCE-196955\"], \"rhea\" => Any[\"13890\", \"13889\", \"13891\", \"13892\"], \"sabiork\" => Any[\"718\"], \"sbo\" => \"SBO:0000176\", \"seed.reaction\" => Any[\"rxn00121\"]…))\n",
       " JSON.Object(\"id\" => \"CAMPt\", \"name\" => \"CAMP transport (ATP-dependent)\", \"metabolites\" => JSON.Object{String, Any}(\"adp_c\" => 1.0, \"atp_c\" => -1.0, \"camp_c\" => -1.0, \"camp_e\" => 1.0, \"h2o_c\" => -1.0, \"h_c\" => 1.0, \"pi_c\" => 1.0), \"lower_bound\" => 0.0, \"upper_bound\" => 1000.0, \"gene_reaction_rule\" => \"Abcc4_1_e\", \"subsystem\" => \"Transport, Extracellular\", \"notes\" => JSON.Object{String, Any}(\"original_bigg_ids\" => Any[\"CAMPt\"]), \"annotation\" => JSON.Object{String, Any}(\"bigg.reaction\" => Any[\"CAMPt\"], \"metanetx.reaction\" => Any[\"MNXR96448\"], \"sbo\" => \"SBO:0000185\"))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "top_k_reactions_softmax = let\n",
    "\n",
    "    # initialize -\n",
    "    k = 10; # how many top reactions do we want?\n",
    "    transformed_v₁ = softmax(v₁); # apply softmax to the eigenvector\n",
    "    sorted_indices = sortperm(transformed_v₁, rev = true); # get sorted indices (largest to smallest)\n",
    "    top_k_indices = sorted_indices[1:k]; # get top k indices\n",
    "    top_k_reactions_softmax = [model[\"reactions\"][i] for i ∈ top_k_indices]; # get top k reactions\n",
    "\n",
    "    # what is the cumulative contribution of these top k reactions?\n",
    "    cumulative_contribution = sum(transformed_v₁[top_k_indices]);\n",
    "    println(\"Cumulative contribution of top $k reactions (softmax): $(round(cumulative_contribution*100, digits=2))%\");\n",
    "\n",
    "    top_k_reactions_softmax # return top k reactions\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95eabb9b",
   "metadata": {},
   "source": [
    "__Interesting!__ Some of the reactions that show up in the top `k` list using the softmax transformation are similar to the ones we found before, but some are different. Another interesting point is the cumulative sum of the top `k` components (much less for the softmax transformation). Hmmm. Mysteries for another day!\n",
    "\n",
    "___"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a680e1b8",
   "metadata": {},
   "source": [
    "## Summary\n",
    "The power iteration method efficiently computes the dominant eigenvalue and eigenvector of a covariance matrix, revealing which reactions contribute most to the structure of a metabolic network.\n",
    "\n",
    "> __Key Takeaways__:\n",
    "> \n",
    "> * **Covariance matrices enable eigendecomposition of non-square matrices:** By computing the covariance matrix of the stoichiometric matrix columns, we transform a non-square matrix into a square, symmetric matrix suitable for eigendecomposition.\n",
    "> * **Power iteration converges to the dominant eigenpair:** The iterative approach provides a computationally efficient alternative to direct methods, particularly useful for large sparse matrices encountered in genome-scale metabolic networks.\n",
    "> * **Eigenvector components identify important network reactions:** Scaling eigenvector components reveals which reactions have the strongest influence on network behavior, providing biological insights into metabolic pathway organization.\n",
    "\n",
    "Eigenvector analysis of metabolic networks reveals the underlying structure that drives biochemical transformations in living systems.\n",
    "___"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.12.3",
   "language": "julia",
   "name": "julia-1.12"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}